libvast/test/module.cpp

```diff
@@ -0,0 +1,688 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2022 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#define SUITE module
+
+#include "vast/module.hpp"
+
+#include "vast/aliases.hpp"
+#include "vast/concept/convertible/data.hpp"
+#include "vast/data.hpp"
+#include "vast/legacy_type.hpp"
+#include "vast/test/test.hpp"
+#include "vast/type.hpp"
+
+#include <caf/error.hpp>
+#include <caf/sum_type.hpp>
+#include <caf/test/dsl.hpp>
+#include <fmt/format.h>
+
+#include <optional>
+#include <string_view>
+
+using namespace vast;
+
+caf::expected<type>
+to_type(const std::vector<type>& known_types, const data& declaration,
+        std::string_view name = "");
+
+caf::expected<record_type>
+to_record(const std::vector<type>& known_types,
+          const std::vector<vast::data>& record_list) {
+  if (record_list.empty())
+    return caf::make_error(ec::parse_error, "parses an empty record");
+  auto record_fields = std::vector<record_type::field_view>{};
+  for (const auto& record_value : record_list) {
+    const auto* record_record_ptr = caf::get_if<record>(&record_value);
+    if (record_record_ptr == nullptr)
+      return caf::make_error(ec::parse_error, "parses an record with invalid "
+                                              "format");
+    const auto& record_record = *record_record_ptr;
+    if (record_record.size() != 1)
+      return caf::make_error(ec::parse_error, "parses an record field with an "
+                                              "invalid format");
+    auto type_or_error = to_type(known_types, record_record.begin()->second);
+    if (!type_or_error)
+      return type_or_error.error();
+    record_fields.emplace_back(record_record.begin()->first, *type_or_error);
+  }
+  return record_type{record_fields};
+}
+
+type get_known_type(const std::vector<type>& known_types,
+                    std::string_view name) {
+  for (const auto& known_type : known_types) {
+    if (name == known_type.name()) {
+      return known_type;
+    }
+  }
+  return {}; // none type
+}
+
+std::vector<std::string> reserved_names
+  = {"bool",   "integer", "count",   "real",   "duration", "time",
+     "string", "pattern", "address", "subnet", "type",     "attributes",
+     "enum",   "list",    "map",     "record", "base",     "implant",
+     "extend", "fields",  "key",     "value"};
+
+caf::expected<type> to_type(const std::vector<type>& known_types,
+                            const data& declaration, std::string_view name) {
```

Comment:
The old parser checked whether the type name conformed to a given set of naming rules. Should we still enforce this or are we just more lenient now because the YAML grammar is more forgiving?

Reply:
This is not so much about YAML. It's more so related to the query side, e.g., our extractor syntax.

Reply:
On the extractor side we can accept everything if the user escapes it. We have no restrictions for it.

Note that this is about _type_ names, not about the names of _fields_ in record types. For fields we absolutely do not need to restrict anything, because they need to be able to match anything the user possibly wants to import.

Reply:
Okay, makes sense. I think we can start as open as possible. I would perhaps avoid `.` and `:`, because these have special meanings in a full-qualified field and in YAML itself.

Reply:
The `type suricata.component.common = record {...}` in the current schema file can be translated to the new YAML module syntax multiple ways:

- Option 1:
```yaml
module: suricata
types: 
  component:
    record:
    - common:
        record:
        - ...
```

- Option 2:
```yaml
module: suricata
types: 
  component:
    record:
    - common: common_record

  common_record:
    record:
    - ...
```

- Option 3:
```yaml
module: suricata
types: 
  component.common:
    record:
      ...
```
I think we are aiming for Option 1 or Option 2. Option 3 would require `.` in the type name.

The `:` may be allowed though if it gets through the YAML parser and gets into a type name, we may just use it:

```yaml
module: foo
types: 
  "app:proto":
    type: string
```


The new commit prevents both `.` and `:` in type names for now.

Reply:
I still don't understand why either of `.` or `:` would cause an issue in type names and don't understand why we'd restrict them at all.

As far as my understanding goes handling escaping purely needs to be done on the extractor side because that receives a flat representation like `suricata.component.common.timestamp`, but the extraction logic already can disambiguate this because the type itself is not flattened in any way.

From the given examples I think we're aiming for option 3. Option 1 and 2 actually cannot work, because Suricata looks like this:

```json5
{
  "event_type": "flow",
  //...
}
```

Rather than like this, which would match the options 1 and 2:

```json5
{
  "common": {
    "event_type": "flow",
    // ...
  },
  // ...
}
```

Reply:
This clears the question. We must allow `.` in type names and there is no reason to restrict other characters.

I have left a comment in the code to reflect this and added a sentence to the YAML Modules doc.

The old parser used:

```
      = ("type" >> skp >> parsers::identifier >> skp >> '=' >> skp >> tp)
```

and `parsers::identifier` is:

```
constexpr inline auto identifier_char = (alnum | ch<'_'> | ch<'.'>);
constexpr inline auto identifier = +identifier_char;
```

The old schema parser allowed only alphanumeric characters the `_` and the `.`. The new parser will not impose these limitations.